File: proj08/java/TemperatureTester.java
import java.util.Scanner;
import java.util.InputMismatchException;

/*
 * Driver for the Temperature type.
 * Author: David Reidsma
 * Date: 04/15/2020
 */
class TemperatureTester {
    public static void main(String[] args) {
	System.out.println("Temperature examples:");
	System.out.println("\t20.5 C");
	System.out.println("\t10 K");
	System.out.println("\t35.1 F");
	System.out.print("Enter a start temperature: ");
	Temperature start = Temperature.readTemp();
	while (start == null) {
	    System.out.print("Invalid input, try again: ");
	    start = Temperature.readTemp();
	}
	
	System.out.print("Enter a stop temperature: ");
	Temperature stop = Temperature.readTemp();
	while (stop == null) {
	    System.out.print("Invalid input, try again: ");
	    stop = Temperature.readTemp();
	}
        
	Scanner scanner = new Scanner(System.in);
	String stepString = "";
	float step = 0;
	boolean hasStep = false;
	int i = 0;
	while (!hasStep) {
	    if (i == 0)
		System.out.print("Enter a step amount of degrees (in the same scale as your start temperature): ");
	    else
		System.out.print("Invalid input, try again: ");
	    try {
		stepString = scanner.nextLine();
		step = Float.parseFloat(stepString);
		hasStep = true;
	    } catch (NumberFormatException e) {	}
	    i++;
	}
	scanner.close();

	if (start.equals(stop)) {
	    System.out.println("Start and stop temperatures are already equal.");
	    return;
	}

	Temperature current = start;
	while (true) {
	    if (start.lessThan(stop) != current.lessThan(stop) && !current.equals(stop))
		break;

	    current.toFahrenheit().print();
	    System.out.print("\t\t");
	    current.toCelsius().print();
	    System.out.print("\t\t");
	    current.toKelvin().print();
	    System.out.println();

	    if (start.lessThan(stop))
		current = current.raise(step);
	    else
		current = current.lower(step);
	}
    }
}

File: proj08/ruby/temperature_tester.rb
require './temperature'

# Driver for the Temperature class.
# Author: David Reidsma
# Date: 04/15/2020
def main
  puts "Temperature examples:"
  puts "\t20.5 C"
  puts "\t10 K"
  puts "\t35.1 F"
  print "Enter a start temperature: "
  start, error = Temperatures::Temperature.read_temp
  while error
    print "Invalid input, try again: "
    start, error = Temperatures::Temperature.read_temp
  end

  print "Enter a stop temperature: "
  stop, error = Temperatures::Temperature.read_temp
  while error
    print "Invalid input, try again: "
    stop, error = Temperatures::Temperature.read_temp
  end
  
  print "Enter a step amount of degrees (in the same scale as your start temperature): "
  step_string = gets.chomp
  step = 0.0
  while not Temperatures::Temperature.is_numeric(step_string)
    print "Invalid input, try again: "
    step_string = gets.chomp
  end
  step = step_string.to_f

  if start.equals(stop)
    puts("Start and stop temperatures are already equal.")
    return
  end
  
  current = start
  while true
    if start.less_than(stop) != current.less_than(stop) && !current.equals(stop)
      break
    end

    current.to_fahrenheit.print_temp
    print "\t\t"
    current.to_celsius.print_temp
    print "\t\t"
    current.to_kelvin.print_temp
    puts ""

    if start.less_than(stop)
      current = current.raise(step)
    else
      current = current.lower(step)
    end
  end
end

main
